authorization/token/service/token_service.go

```diff
@@ -0,0 +1,326 @@
+	if loadedToken != nil && tokenClaims.Permissions != nil {
+		for _, tokenPermission := range *tokenClaims.Permissions {
+			if *tokenPermission.ResourceSetID == resourceID {
+				resourceExistsInToken = true
+			}
+		}
+	}
+
+	if loadedToken != nil {
+		// Confirm that the token belongs to the current identity
+		if loadedToken.IdentityID != identity.ID {
+			return nil, errors.NewUnauthorizedError("invalid token for identity")
+		}
+
+		// If the token exists and its status is valid, return an empty string
+		if loadedToken.Valid() && resourceExistsInToken {
+			return nil, nil
+		}
+
+		// We now process the various token status codes in order of priority, starting with DEPROVISIONED
+		if loadedToken.HasStatus(tokenPkg.TOKEN_STATUS_DEPROVISIONED) {
+			return nil, errors.NewUnauthorizedErrorWithCode("token deprovisioned", errors.UNAUTHORIZED_CODE_TOKEN_DEPROVISIONED)
+		}
+
+		// If the token has been revoked or the user is logged out, we respond in the same way
+		if loadedToken.HasStatus(tokenPkg.TOKEN_STATUS_REVOKED) || loadedToken.HasStatus(tokenPkg.TOKEN_STATUS_LOGGED_OUT) {
+			return nil, errors.NewUnauthorizedErrorWithCode("token revoked or logged out", errors.UNAUTHORIZED_CODE_TOKEN_REVOKED)
+		}
+
+		// If the token is stale, yet the resource exists in the token
+		// we can re-evaluate its privileges to determine whether they have changed.
+		// If the privileges are unchanged, then reset the token status
+		if loadedToken.HasStatus(tokenPkg.TOKEN_STATUS_STALE) && resourceExistsInToken {
+			// Query for all of the token's privileges
+			privileges, err := s.Repositories().TokenRepository().ListPrivileges(ctx, tokenID)
+			if err != nil {
+				return nil, errors.NewInternalError(ctx, err)
+			}
+
+			scopesChanged := false
+			permissionsChanged := false
+
+			// If the number of permissions in the token is different to what is stored in the database, then
+			// something has definitely changed
+			if len(privileges) != len(*tokenClaims.Permissions) {
+				permissionsChanged = true
+			} else {
+				// Compare the scopes to those contained in the current token
+				for _, tokenPermission := range *tokenClaims.Permissions {
+					// Retrieve the up to date scopes for the resource
+					privilegeCache, err := s.Services().PrivilegeCacheService().CachedPrivileges(ctx, identity.ID, *tokenPermission.ResourceSetID)
+					if err != nil {
+						return nil, errors.NewInternalError(ctx, err)
+					}
+
+					// Compare the scopes of the resource with the scopes in the token
+					scopes := privilegeCache.ScopesAsArray()
+					if !s.scopesEquivalent(tokenPermission.Scopes, scopes) {
+						scopesChanged = true
+						break
+					}
+
+					resourceFound := false
+
+					// Also confirm that the resources in the token's permissions match those stored in the database
+					for _, priv := range privileges {
+						if priv.ResourceID == *tokenPermission.ResourceSetID {
+							resourceFound = true
+						}
+					}
+
+					if !resourceFound {
```

Comment:
Curious, when could this possibly happen?
Maybe a test for this could look like:

1. add a new RPT token for an identity, relate them with the privilege_cache' relation.
2. generate RPT
3. add more privileges for the corresponding token_id in the privilege_cache.

Reply:
This scenario is more for when a resource is deleted. I don't think we need to worry about additional privileges being added manually to the database, if someone owns our database then there's not much we can do about that in code. :)